From 4aeb2409c2c77cfe8ad1d85d66290116f3d4d951 Mon Sep 17 00:00:00 2001
From: sai surya teja <sssteja@habana.ai>
Date: Wed, 19 Apr 2023 04:03:39 +0300
Subject: [PATCH] [SW-140835] sec: gaudi2: mode based dynamic image sram opening

This patch aligns erom and preboot based on the following patch:
https://gerrit.habana-labs.com/#/c/197942/

This patch add support to open sram image region based on the
access mode requested. In gaudi2, RO and RW modes can be supported.

Change-Id: Ic34a7523cde9287de26d50bd886dbe91ec8704f9
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index d1f3684..61fb48a 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -1247,9 +1247,25 @@
 	cfg = &preboot_sec_dyn_cfg[IMG_END_TO_SRAM_END];
 	hl_secure_dyn_rr_cfg(cfg->rr, cfg->min, cfg->max);
 
-	/* TODO: Check mode based opening possibility for gaudi2 */
 	/* Finally open the image memory space */
-	hl_secure_dyn_rr_cfg(preboot_sec_dyn_cfg[IMAGE].rr, 0, 0);
+	cfg = &preboot_sec_dyn_cfg[IMAGE];
+	switch (mode) {
+	case ACCESS_RO:
+		/* Open IMAGE region in read only mode */
+		hl_secure_dyn_rr_cfg(HL_SEC_HBW_SRAM_RO, cfg->min, cfg->max);
+		break;
+	case ACCESS_RW:
+		/* Open IMAGE region read-only RR if closed already */
+		hl_secure_dyn_rr_cfg(HL_SEC_HBW_SRAM_RO, 0, 0);
+		break;
+	case ACCESS_WR: /* write-only not possible in gaudi2 */
+	default:
+		preboot_puts_hex("dynamic_image_sram_open with invalid mode: ", mode, false);
+		return;
+	}
+
+	/* Open IMAGE region read-write RR */
+	hl_secure_dyn_rr_cfg(cfg->rr, 0, 0);
 }
 
 void preboot_security_dynamic_img_sram_close(void)
